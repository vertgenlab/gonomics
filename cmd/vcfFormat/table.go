package main

import (
	"fmt"
	"github.com/vertgenlab/gonomics/exception"
	"github.com/vertgenlab/gonomics/vcf"
	"io"
	"log"
	"strconv"
	"strings"
)

func writeTableHeader(outfile io.Writer, header vcf.Header, maxAlts int) (infoOrder []vcf.InfoHeader, formatOrder []vcf.FormatHeader) {
	s := new(strings.Builder)
	if len(header.Text) == 0 {
		log.Fatal("ERROR: no vcf header found. Must have well-formed header to use -tsv")
	}

	// write common header components
	s.WriteString("Chromosome,Position,ID,Reference")
	if maxAlts == 1 {
		s.WriteString(",Alternate")
	} else {
		for i := 0; i < maxAlts; i++ {
			s.WriteString(",Alternate_" + fmt.Sprintf("%d", i))
		}
	}
	s.WriteString(",Quality,Filter")

	// write INFO field
	var numFields int
	for key, val := range header.Info {
		numFields = numberOfFields(maxAlts, val.Key)
		if numFields == 1 {
			s.WriteString("," + key)
		} else {
			for i := 0; i < numFields; i++ {
				s.WriteString("," + key + fmt.Sprintf("_%d", i))
			}
		}
		infoOrder = append(infoOrder, val)
	}

	// get formatOrder
	for _, val := range header.Format {
		formatOrder = append(formatOrder, val)
	}

	// get sample order
	sampleOrder := make([]string, len(header.Samples))
	for key, val := range header.Samples {
		sampleOrder[val] = key
	}

	// write format per sample
	for _, fmtHeader := range formatOrder {
		for _, sample := range sampleOrder {
			numFields = numberOfFields(maxAlts, fmtHeader.Key)
			if numFields == 1 {
				s.WriteString("," + fmtHeader.Id + "_" + sample)
			} else {
				for i := 0; i < numFields; i++ {
					s.WriteString("," + fmtHeader.Id + "_" + sample + fmt.Sprintf("_%d", i))
				}
			}
		}
	}

	_, err := fmt.Fprintln(outfile, s.String())
	exception.PanicOnErr(err)
	return
}

func writeAsTable(s *strings.Builder, outfile io.Writer, v vcf.Vcf, header vcf.Header, infoOrder []vcf.InfoHeader, formatOrder []vcf.FormatHeader, maxAlts int) {
	s.Reset()

	// write basic data
	s.WriteString(fmt.Sprintf("%s,%d,%s,%s,%s",
		v.Chr, v.Pos, v.Id, v.Ref, strings.Join(v.Alt, ",")))
	for i := len(v.Alt); i < maxAlts; i++ {
		s.WriteString(",")
	}
	s.WriteString(fmt.Sprintf(",%g,%s", v.Qual, v.Filter))

	// write info data
	v = vcf.ParseInfo(v, header)
	for i := range infoOrder {
		writeData(s, v, infoOrder[i].Key, numberOfFields(maxAlts, infoOrder[i].Key), 1)
	}

	// write format data
	v = vcf.ParseFormat(v, header)
	for i := range formatOrder {
		writeData(s, v, formatOrder[i].Key, numberOfFields(maxAlts, formatOrder[i].Key), len(v.Samples))
	}

	_, err := fmt.Fprintln(outfile, s.String())
	exception.PanicOnErr(err)
}

// getMaxAltCount reads through the input vcf file to determine the maximum number of alternate alleles present
func getMaxAltCount(infile string) int {
	var maxAlts int
	records, _ := vcf.GoReadToChan(infile)
	for v := range records {
		if len(v.Alt) > maxAlts {
			maxAlts = len(v.Alt)
		}
	}
	return maxAlts
}

func numberOfFields(maxAlts int, k vcf.Key) int {
	switch k.Number {
	case "A": // == num alt alleles
		return maxAlts

	case "R": // == num ref + alt alleles
		return maxAlts + 1

	case "G": // one value for each possible genotype
		return 1 //TODO once parser is updated

	case ".": // wildcard. they never make it easy do they...
		return 1

	default:
		num, err := strconv.Atoi(k.Number)
		if err != nil {
			log.Panicf("'%s' is not a valid Number for header info", k.Number)
		}
		return num
	}
}

func writeData(s *strings.Builder, v vcf.Vcf, key vcf.Key, numberOfFieldsPerSample int, repeats int) {
	var innerFieldsWritten, fieldsWritten, i, j int

	switch key.DataType {
	case vcf.Integer:
		ints, found := vcf.QueryInt(v, key)
		if !found {
			break
		}
		for i = range ints { // per sample data
			for j = range ints[i] { // per field data
				s.WriteString(fmt.Sprintf(",%d", ints[i][j]))
				fieldsWritten++
				innerFieldsWritten++
			}
			for j = innerFieldsWritten; j < numberOfFieldsPerSample; j++ {
				s.WriteString(",")
				fieldsWritten++
			}
			innerFieldsWritten = 0
		}

	case vcf.Float:
		flts, found := vcf.QueryFloat(v, key)
		if !found {
			break
		}
		for i = range flts { // per sample data
			for j = range flts[i] { // per field data
				s.WriteString(fmt.Sprintf(",%g", flts[i][j]))
				fieldsWritten++
				innerFieldsWritten++
			}
			for j = innerFieldsWritten; j < numberOfFieldsPerSample; j++ {
				s.WriteString(",")
				fieldsWritten++
			}
			innerFieldsWritten = 0
		}

	case vcf.String:
		strs, found := vcf.QueryString(v, key)
		if !found {
			break
		}
		for i = range strs { // per sample data
			for j = range strs[i] { // per field data
				s.WriteString(fmt.Sprintf(",%s", strs[i][j]))
				fieldsWritten++
				innerFieldsWritten++
			}
			for j = innerFieldsWritten; j < numberOfFieldsPerSample; j++ {
				s.WriteString(",")
				fieldsWritten++
			}
			innerFieldsWritten = 0
		}

	case vcf.Character:
		chars, found := vcf.QueryRune(v, key)
		if !found {
			break
		}
		for i = range chars { // per sample data
			for j = range chars[i] { // per field data
				s.WriteString(fmt.Sprintf(",%c", chars[i][j]))
				fieldsWritten++
				innerFieldsWritten++
			}
			for j = innerFieldsWritten; j < numberOfFieldsPerSample; j++ {
				s.WriteString(",")
				fieldsWritten++
			}
			innerFieldsWritten = 0
		}

	case vcf.Flag:
		found := vcf.QueryFlag(v, key)
		if found {
			s.WriteString(",TRUE")
		} else {
			s.WriteString(",FALSE")
		}
		fieldsWritten++
		innerFieldsWritten++
		for j = innerFieldsWritten; j < numberOfFieldsPerSample; j++ {
			s.WriteString(",")
			fieldsWritten++
		}
		innerFieldsWritten = 0
	}

	for j = fieldsWritten; j < numberOfFieldsPerSample*repeats; j++ {
		s.WriteString(",")
	}
}
