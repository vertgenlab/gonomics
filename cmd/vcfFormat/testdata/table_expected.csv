Chromosome,Position,ID,Reference,Alternate_0,Alternate_1,Quality,Filter,MaxPopAF,Consequence,Gene,Transcript,ProteinEffect,GT_S2007.recal,GT_S2032.recal,GT_S4322.recal,GT_S4388.recal,GT_S4390.recal,DP_S2007.recal,DP_S2032.recal,DP_S4322.recal,DP_S4388.recal,DP_S4390.recal,AD_S2007.recal_0,AD_S2007.recal_1,AD_S2007.recal_2,AD_S2032.recal_0,AD_S2032.recal_1,AD_S2032.recal_2,AD_S4322.recal_0,AD_S4322.recal_1,AD_S4322.recal_2,AD_S4388.recal_0,AD_S4388.recal_1,AD_S4388.recal_2,AD_S4390.recal_0,AD_S4390.recal_1,AD_S4390.recal_2,PV_S2007.recal_0,PV_S2007.recal_1,PV_S2032.recal_0,PV_S2032.recal_1,PV_S4322.recal_0,PV_S4322.recal_1,PV_S4388.recal_0,PV_S4388.recal_1,PV_S4390.recal_0,PV_S4390.recal_1
chr1,14997,.,A,C,T,0,NO_NORMAL,,splice_region_variant,DDX11L1,ENST00000456328,,,,,,,797,111,981,653,982,795,1,1,111,0,0,975,3,0,625,28,0,981,0,0,1,1,1,1,1,1,3e-17,1,1,1
chr1,17606,.,A,G,,0,NO_NORMAL,0.0043,splice_region_variant,DDX11L1,ENST00000456328,,,,,,,1477,111,1560,862,1581,1472,0,,97,12,,1549,0,,854,0,,1574,2,,1,,2e-19,,1,,1,,1,
chr1,69231,.,C,G,,0,NO_NORMAL,,missense_variant,OR4F5,ENST00000641515,HIS47GLN,,,,,,1095,146,1362,767,1846,1082,11,,146,0,,1360,0,,767,0,,1837,8,,0.0005,,1,,1,,1,,1,
chr1,69356,.,G,T,,0,NO_NORMAL,,missense_variant,OR4F5,ENST00000641515,GLY89VAL,,,,,,494,137,607,566,1566,494,0,,137,0,,600,7,,566,0,,1566,0,,1,,1,,6e-06,,1,,1,
chr1,69369,.,G,A,,0,NO_NORMAL,0.0005,synonymous_variant,OR4F5,ENST00000641515,GLN93GLN,,,,,,399,135,477,582,1645,399,0,,134,0,,412,64,,582,0,,1645,0,,1,,1,,1e-55,,1,,1,
chr1,69437,.,A,G,,0,NO_NORMAL,,missense_variant,OR4F5,ENST00000641515,TYR116CYS,,,,,,62,188,63,803,2508,61,0,,188,0,,61,2,,803,0,,2505,0,,1,,1,,0.0003,,1,,1,
chr1,137334,.,C,T,,0,NO_NORMAL,0.0036,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,56,0,62,0,77,56,0,,0,0,,62,0,,0,0,,64,13,,1,,1,,1,,1,,3e-06,
chr1,137377,.,G,T,,0,NO_NORMAL,0.0091,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,91,0,107,73,123,91,0,,0,0,,107,0,,73,0,,116,7,,1,,1,,1,,1,,0.0003,
chr1,137411,.,G,A,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,117,0,137,105,179,117,0,,0,0,,125,12,,105,0,,179,0,,1,,1,,5e-08,,1,,1,
chr1,137486,.,A,G,,0,NO_NORMAL,0.0015,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,168,0,206,128,282,168,0,,0,0,,204,0,,115,11,,281,0,,1,,1,,1,,2e-09,,1,
chr1,137487,.,A,C,,0,NO_NORMAL,0.0015,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,169,0,206,129,284,169,0,,0,0,,203,3,,116,13,,283,1,,1,,1,,1,,5e-08,,1,
chr1,137552,.,T,C,,0,NO_NORMAL,0.0014,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,177,0,185,119,302,176,0,,0,0,,184,0,,119,0,,270,30,,1,,1,,1,,1,,2e-13,
chr1,137560,.,GC,G,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,183,0,178,115,295,183,0,,0,0,,178,0,,115,0,,268,27,,1,,1,,1,,1,,3e-12,
chr1,137987,.,G,A,,0,NO_NORMAL,0.00014,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,77,0,79,69,126,76,0,,0,0,,79,0,,64,5,,126,0,,1,,1,,1,,0.0003,,1,
chr1,138864,.,G,A,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,622,63,589,542,1229,621,0,,61,2,,586,0,,542,0,,1227,0,,1,,0.0004,,1,,1,,1,
chr1,138875,.,G,A,,0,NO_NORMAL,0.00017,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,646,65,598,572,1267,646,0,,64,0,,596,0,,563,9,,1263,1,,1,,1,,1,,2e-06,,1,
chr1,138909,.,C,T,,0,NO_NORMAL,0.0089,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,711,58,649,651,1415,711,0,,58,0,,622,27,,650,0,,1413,0,,1,,1,,1e-20,,1,,1,
chr1,138955,.,G,T,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,814,90,740,811,1792,814,0,,90,0,,740,0,,800,11,,1792,0,,1,,1,,1,,1e-08,,1,
chr1,138980,.,G,T,,0,NO_NORMAL,0.00091,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,832,102,764,912,1989,801,31,,102,0,,734,29,,899,13,,1981,8,,1e-05,,1,,2e-05,,1,,1,
chr1,139054,.,T,C,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,855,135,686,1034,2296,853,0,,133,2,,686,0,,1033,0,,2293,0,,1,,0.0007,,1,,1,,1,
chr1,139115,.,G,T,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,977,110,810,1098,2321,976,1,,107,3,,810,0,,1097,1,,2320,1,,1,,0.0002,,1,,1,,1,
chr1,139132,.,G,T,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,989,105,874,1079,2327,989,0,,102,3,,873,1,,1074,3,,2323,2,,1,,0.0006,,1,,1,,1,
chr1,139191,.,C,A,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,1056,101,1163,1190,2295,1055,1,,98,3,,1161,0,,1189,1,,2292,3,,1,,0.0003,,1,,1,,1,
chr1,139200,.,G,A,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,1076,94,1209,1184,2321,1076,0,,93,0,,1189,19,,1172,11,,2313,7,,1,,1,,3e-05,,1,,1,
chr1,139208,.,G,A,,0,NO_NORMAL,0.00023,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,1074,91,1225,1171,2323,1074,0,,91,0,,1208,17,,1170,0,,2322,0,,1,,1,,2e-12,,1,,1,
chr1,139356,.,T,C,,0,NO_NORMAL,0.0039,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,792,72,754,749,1780,789,1,,72,0,,733,18,,749,0,,1776,1,,1,,1,,5e-12,,1,,1,
chr1,139363,.,T,C,,0,NO_NORMAL,0.0019,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,767,64,730,717,1720,760,1,,64,0,,719,10,,716,0,,1717,1,,1,,1,,2e-06,,1,,1,
chr1,139415,.,G,T,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,632,67,577,594,1463,632,0,,67,0,,577,0,,588,6,,1461,0,,1,,1,,1,,3e-05,,1,
chr1,139491,.,G,A,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,593,68,586,553,1401,589,2,,64,4,,586,0,,552,0,,1397,0,,1,,3e-06,,1,,1,,1,
chr1,139666,.,T,G,,0,NO_NORMAL,,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,280,0,328,272,747,271,9,,0,0,,315,13,,250,22,,730,16,,1,,1,,0.4,,0.0001,,1,
chr1,139685,.,G,A,,0,NO_NORMAL,0.0011,2KB_upstream_variant,CICP27,ENST00000442987,,,,,,,220,0,278,214,645,219,0,,0,0,,277,0,,214,0,,625,15,,1,,1,,1,,1,,1e-05,
chr1,177463,.,A,G,,0,NO_NORMAL,,upstream_gene_variant,AL627309.5,ENST00000466557,,,,,,,391,0,371,224,446,391,0,,0,0,,370,0,,224,0,,409,37,,1,,1,,1,,1,,6e-20,
chr1,182441,.,T,A,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,76,0,60,0,93,68,8,,0,0,,48,12,,0,0,,93,0,,0.3,,1,,0.0009,,1,,1,
chr1,182529,.,G,A,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,303,0,222,153,271,303,0,,0,0,,222,0,,153,0,,263,8,,1,,1,,1,,1,,4e-05,
chr1,182631,.,A,G,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1023,60,869,640,923,1021,1,,60,0,,867,0,,639,0,,790,133,,1,,1,,1,,1,,3e-79,
chr1,182640,.,C,A,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1087,62,943,700,971,1082,5,,58,4,,940,2,,700,0,,964,7,,1,,0.0002,,1,,1,,1,
chr1,182668,.,T,C,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1135,62,995,737,977,1133,2,,58,4,,991,3,,737,0,,973,4,,1,,4e-05,,1,,1,,1,
chr1,182678,.,A,G,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1167,55,1009,753,971,1164,2,,52,3,,1006,2,,753,0,,967,4,,1,,0.0004,,1,,1,,1,
chr1,182686,.,A,G,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1183,51,1004,747,955,975,207,,51,0,,844,154,,600,144,,952,0,,1e-08,,1,,0.004,,1e-08,,1,
chr1,182688,.,A,G,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.2,ENST00000624431,,,,,,,1184,50,1009,747,963,1180,2,,47,3,,1005,2,,747,0,,957,4,,1,,0.0003,,1,,1,,1,
chr1,184246,.,T,C,,0,NO_NORMAL,,2KB_downstream_variant,FO538757.2,ENST00000624431,,,,,,,392,0,332,197,315,391,0,,0,0,,332,0,,178,19,,315,0,,1,,1,,1,,3e-16,,1,
chr1,185194,.,G,C,,0,NO_NORMAL,,2KB_downstream_variant,FO538757.2,ENST00000624431,,,,,,,448,0,338,243,314,209,238,,0,0,,219,119,,154,89,,259,55,,3e-17,,1,,1,,1,,1,
chr1,188130,.,A,G,,0,NO_NORMAL,,splice_region_variant,FO538757.2,ENST00000624431,,,,,,,751,61,942,538,882,744,2,,52,9,,941,0,,534,0,,880,0,,1,,1e-14,,1,,1,,1,
chr1,195425,.,C,G,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.1,ENST00000623083,,,,,,,82,0,58,0,97,47,35,,0,0,,37,21,,0,0,,85,12,,0.0005,,1,,0.1,,1,,1,
chr1,195438,.,T,A,,0,NO_NORMAL,,2KB_upstream_variant,FO538757.1,ENST00000623083,,,,,,,74,0,53,0,82,74,0,,0,0,,53,0,,0,0,,74,8,,1,,1,,1,,1,,0.0005,
chr1,199681,.,C,T,,0,NO_NORMAL,,upstream_gene_variant,FO538757.1,ENST00000623083,,,,,,,65,0,76,0,106,62,3,,0,0,,60,16,,0,0,,100,3,,1,,1,,3e-05,,1,,1,
chr1,199869,.,C,T,,0,NO_NORMAL,,upstream_gene_variant,FO538757.1,ENST00000623083,,,,,,,69,0,56,0,107,59,10,,0,0,,56,0,,0,0,,107,0,,3e-06,,1,,1,,1,,1,
chr1,199878,.,T,C,,0,NO_NORMAL,,upstream_gene_variant,FO538757.1,ENST00000623083,,,,,,,67,0,56,0,105,62,2,,0,0,,45,11,,0,0,,103,1,,1,,1,,2e-05,,1,,1,
chr1,200960,.,T,C,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,179,0,247,184,351,179,0,,0,0,,246,0,,184,0,,278,72,,1,,1,,1,,1,,2e-34,
chr1,205549,.,G,C,,0,NO_NORMAL,,TF_binding_site_variant,,,,,,,,,118,0,148,77,130,118,0,,0,0,,135,13,,77,0,,130,0,,1,,1,,2e-07,,1,,1,
chr1,205611,.,C,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,269,0,367,201,298,226,43,,0,0,,337,30,,184,17,,270,28,,0.0007,,1,,1,,1,,1,
chr1,205688,.,G,A,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,631,61,644,477,664,523,107,,60,1,,642,1,,443,34,,547,117,,3e-09,,1,,1,,1,,2e-11,
chr1,205748,.,G,C,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,817,77,685,616,874,650,164,,73,4,,683,0,,553,61,,669,204,,2e-08,,1,,1,,1,,4e-19,
chr1,205757,.,C,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,833,75,705,641,878,586,245,,51,24,,449,256,,463,173,,598,277,,1,,0.5,,0.0005,,1,,0.4,
chr1,205818,.,C,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,799,89,701,607,862,553,245,,60,29,,406,293,,424,183,,562,298,,1,,1,,1e-06,,1,,0.4,
chr1,205855,.,T,C,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,650,62,610,454,714,445,202,,42,20,,357,252,,308,146,,473,240,,1,,1,,4e-05,,1,,1,
chr1,205873,.,G,A,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,561,56,539,388,627,560,0,,54,2,,537,0,,387,0,,627,0,,1,,0.0007,,1,,1,,1,
chr1,205882,.,T,C,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,538,52,516,366,589,296,241,,28,24,,307,208,,174,191,,265,322,,1,,1,,1,,0.04,,5e-05,
chr1,205918,.,A,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,446,0,435,282,496,444,1,,0,0,,431,1,,266,16,,496,0,,1,,1,,1,,4e-11,,1,
chr1,205931,.,A,G,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,386,0,389,236,439,350,32,,0,0,,387,1,,223,12,,384,54,,0.1,,1,,1,,1,,2e-07,
chr1,205932,.,A,G,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,386,0,386,235,434,355,29,,0,0,,383,0,,223,12,,378,54,,0.2,,1,,1,,1,,2e-08,
chr1,205980,.,C,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,282,0,269,163,307,282,0,,0,0,,223,45,,163,0,,307,0,,1,,1,,5e-28,,1,,1,
chr1,206076,.,TC,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,422,0,377,259,445,420,0,,0,0,,377,0,,252,6,,445,0,,1,,1,,1,,2e-05,,1,
chr1,206154,.,C,T,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,549,0,539,354,621,548,0,,0,0,,539,0,,349,4,,619,0,,1,,1,,1,,0.0009,,1,
chr1,206436,.,C,G,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,56,0,0,0,57,39,17,,0,0,,0,0,,0,0,,55,2,,0.0001,,1,,1,,1,,1,
chr1,206443,.,G,A,,0,NO_NORMAL,,intergenic_variant,,,,,,,,,50,0,0,0,54,35,15,,0,0,,0,0,,0,0,,52,2,,0.0003,,1,,1,,1,,1,
chr1,450790,.,C,A,,0,NO_NORMAL,,missense_variant,AL732372.2,ENST00000455207,,,,,,,693,0,958,586,805,692,1,,0,0,,958,0,,579,7,,805,0,,1,,1,,1,,6e-05,,1,
chr1,450942,.,A,G,,0,NO_NORMAL,,missense_variant,AL732372.2,ENST00000455207,,,,,,,894,59,1384,766,1156,893,1,,50,9,,1381,1,,766,0,,1154,0,,1,,5e-16,,1,,1,,1,
chr1,451360,.,C,T,,0,NO_NORMAL,,missense_variant,AL732372.2,ENST00000455207,,,,,,,950,0,1137,335,837,948,0,,0,0,,1111,23,,307,27,,816,20,,1,,1,,1,,2e-10,,0.3,
chr1,451466,.,T,G,,0,NO_NORMAL,,missense_variant,AL732372.2,ENST00000455207,,,,,,,1067,55,1230,568,1088,1066,0,,53,2,,1229,1,,567,0,,1085,0,,1,,0.0005,,1,,1,,1,
chr1,451475,.,G,A,,0,NO_NORMAL,,synonymous_variant,AL732372.2,ENST00000455207,,,,,,,1059,56,1225,607,1114,1056,0,,56,0,,1132,92,,604,3,,1113,0,,1,,1,,6e-45,,1,,1,
chr1,451532,.,C,T,,0,NO_NORMAL,,synonymous_variant,AL732372.2,ENST00000455207,,,,,,,958,60,1166,634,1097,956,0,,52,8,,1164,0,,634,0,,1096,0,,1,,2e-15,,1,,1,,1,
chr1,451584,.,G,A,,0,NO_NORMAL,,missense_variant,AL732372.2,ENST00000455207,,,,,,,737,51,922,505,843,734,0,,51,0,,853,68,,504,0,,841,2,,1,,1,,8e-34,,1,,1,
chr1,492762,.,GC,G,,0,NO_NORMAL,,splice_region_variant,AL732372.2,ENST00000455464,,,,,,,95,0,140,88,122,95,0,,0,0,,139,0,,88,0,,114,8,,1,,1,,1,,1,,3e-05,
chr1,492769,.,G,A,,0,NO_NORMAL,,splice_region_variant,AL732372.2,ENST00000455464,,,,,,,94,0,141,87,126,94,0,,0,0,,141,0,,82,5,,126,0,,1,,1,,1,,0.0003,,1,
chr1,498046,.,C,T,,0,NO_NORMAL,,splice_donor_variant,AL732372.2,ENST00000455464,,,,,,,105,0,140,86,176,36,68,,0,0,,66,73,,56,30,,134,42,,1e-07,,1,,0.003,,1,,1,
chr1,685802,.,C,A,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,852,59,1232,703,1032,852,0,,57,2,,1232,0,,702,0,,1031,1,,1,,0.0007,,1,,1,,1,
chr1,685883,.,A,T,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,960,70,1462,785,1285,960,0,,68,2,,1462,0,,784,1,,1285,0,,1,,0.0007,,1,,1,,1,
chr1,685918,.,A,G,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,889,53,1284,713,1183,886,0,,42,10,,1280,0,,711,1,,1182,0,,1,,6e-19,,1,,1,,1,
chr1,686174,.,GC,G,,0,NO_NORMAL,,frameshift_variant,AL669831.3,ENST00000419394,,,,,,,707,0,1048,539,971,704,0,,0,0,,1042,0,,533,6,,969,0,,1,,1,,1,,2e-05,,1,
chr1,686266,.,A,C,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,920,59,1114,434,975,917,0,,40,19,,1109,1,,432,0,,975,0,,1,,2e-34,,1,,1,,1,
chr1,686281,.,G,C,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,951,63,1122,408,972,950,0,,60,2,,1119,1,,407,0,,966,0,,1,,0.0009,,1,,1,,1,
chr1,686390,.,G,A,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,1032,56,1160,409,1045,1032,0,,48,8,,1158,0,,408,0,,1044,0,,1,,2e-15,,1,,1,,1,
chr1,686415,.,C,G,,0,NO_NORMAL,,missense_variant,AL669831.3,ENST00000419394,,,,,,,1061,59,1187,455,1129,1061,0,,57,2,,1186,0,,454,0,,1129,0,,1,,0.0002,,1,,1,,1,
chr1,686451,.,G,A,,0,NO_NORMAL,,synonymous_variant,AL669831.3,ENST00000419394,,,,,,,1127,70,1269,539,1228,1127,0,,70,0,,1196,69,,538,0,,1226,0,,1,,1,,2e-37,,1,,1,
chr1,686508,.,C,T,,0,NO_NORMAL,,synonymous_variant,AL669831.3,ENST00000419394,,,,,,,1024,76,1214,585,1180,1021,3,,60,16,,1214,0,,585,0,,1180,0,,1,,4e-26,,1,,1,,1,
